### PR TITLE
feat(env)+docs: ChronoViperWalk-v0 + README — AMD ROCm, Ray, Gymnasium

### PR DESCRIPTION
**Pull request by amd-pratmish**
## Summary

> **PR title (GitHub):** `feat(env)+docs: ChronoViperWalk-v0 + README — AMD ROCm, Ray, Gymnasium`

This pull request **combines** (1) a new **Chrono 10–native** Gymnasium environment and (2) **README** guidance for **version pinning**, optional **PyTorch ROCm + Ray** layouts, and **reviewer-runnable** smoke commands.

### A — New environment: `ChronoViperWalk-v0`

- Headless **`gym.Env`** using **NSC + Bullet** and **`pychrono.robot.Viper`** with Chrono 10 bindings (`ChFramed`, `ChVector3d`, `SetGravitationalAcceleration`, `system.Add`, …).
- Registered in `gym_chrono/__init__.py` as **`ChronoViperWalk-v0`** (`max_episode_steps=100_000`).
- **Observation** `Box(18,)`, **action** `Box(12,)` — same *tensor shapes* as the legacy `quadruped_walk` env for reuse of common policy widths; **reward is not identical** to every published quadruped benchmark (simple forward Δx).
- **`gym_chrono/test/smoke_viper_walk.py`** — minimal `reset` + `step` for CI or manual validation.

### B — README: version matrix + ROCm + Ray

- New sections: **ChronoViperWalk quick start**, **version matrix** (Chrono / Python / gymnasium / numpy / optional rsl-rl-lib / PyTorch), **Ray + PyTorch ROCm** (bootstrap env vars before `import ray`, `runtime_env` for CPU-only sim actors, `ROCR_VISIBLE_DEVICES`, `ray start --num-gpus`).
- Fixed a broken PyChrono install hyperlink in the existing README.

### C — Import hygiene

- **`gym_chrono/envs/__init__.py`** no longer eagerly imports `ChronoBase` (submodules already import from `gym_chrono.envs.ChronoBase` directly).
- **`gym_chrono/envs/legged/__init__.py`** avoids importing `viper_walk` at package import time so `gym.make("ChronoViperWalk-v0")` does not pull unrelated env code first.

---

## What changed (files)

| File | Change |
|------|--------|
| `gym_chrono/envs/legged/viper_walk.py` | **New** — `ChronoViperWalkEnv` + local `_set_chrono_data_path()` (no hard dependency on `pychrono.vehicle` / `pychrono.sensor` via `utils.py` for this env). |
| `gym_chrono/__init__.py` | Register **`ChronoViperWalk-v0`**; normalize quoting for legacy `art_wpts-v0` entry. |
| `gym_chrono/envs/__init__.py` | **Lazy** package docstring (no eager `ChronoBase` import). |
| `gym_chrono/envs/legged/__init__.py` | **Lazy** legged package marker. |
| `gym_chrono/test/smoke_viper_walk.py` | **New** — smoke harness. |
| `README.md` | **Extended** — ChronoViperWalk section, version matrix, ROCm + Ray patterns, runnable shell/Python examples; **link fix** for PyChrono install docs. |

---

### Gymnasium registration

```python
from gymnasium.envs.registration import register

register(
    id="ChronoViperWalk-v0",
    entry_point="gym_chrono.envs.legged.viper_walk:ChronoViperWalkEnv",
    max_episode_steps=100_000,
)
```

### Ray bootstrap before `import ray` (PyTorch ROCm + Ray ≥ 2.45)

```python
import os

os.environ.setdefault("RAY_EXPERIMENTAL_NOSET_HIP_VISIBLE_DEVICES", "1")
os.environ.setdefault("RAY_EXPERIMENTAL_NOSET_ROCR_VISIBLE_DEVICES", "1")

import ray  # noqa: E402
```

### CPU-only Chrono inside a Ray actor (optional pattern)

```python
_CHRONO_CPU_ENV = {
    "env_vars": {
        "HIP_VISIBLE_DEVICES": "-1",
        "ROCR_VISIBLE_DEVICES": "-1",
    }
}


@ray.remote(num_cpus=1, runtime_env=_CHRONO_CPU_ENV)
class ChronoSimActor:
    def __init__(self, seed: int):
        import gymnasium as gym
        import gym_chrono  # noqa: F401

        self._env = gym.make("ChronoViperWalk-v0", render_mode=None)
        self._env.reset(seed=seed)

    def step(self, action):
        return self._env.step(action)
```

---

## Running examples (copy-paste)

**1) Environment smoke (requires PyChrono + `CHRONO_DATA_DIR`):**

```bash
export CHRONO_DATA_DIR="/path/to/chrono/data/"
export PYTHONPATH="/path/to/gym-chrono:$PYTHONPATH"
python3 gym_chrono/test/smoke_viper_walk.py
```

**2) Inline `gym.make` check:**

```bash
python3 - <<'PY'
import gymnasium as gym
import gym_chrono  # noqa: F401 — registration side effect
env = gym.make("ChronoViperWalk-v0", render_mode=None)
obs, _ = env.reset(seed=0)
obs2, r, term, trunc, _ = env.step(env.action_space.sample())
env.close()
print("obs", obs.shape, "->", obs2.shape, "r", r, "done", term or trunc)
PY
```

**3) Optional syntax-only check (no PyChrono import of the env at runtime):**

```bash
python3 -m py_compile gym_chrono/envs/legged/viper_walk.py gym_chrono/test/smoke_viper_walk.py
```

---

### Environment (`ChronoViperWalk-v0`)

- [ ] Install or build **PyChrono** with **`pychrono.robot`** on `PYTHONPATH`.
- [ ] Set **`CHRONO_DATA_DIR`** to a valid Chrono data directory (textures optional; ground texture load is best-effort).
- [ ] `pip install gymnasium numpy` in a clean venv.
- [ ] Run **`python3 gym_chrono/test/smoke_viper_walk.py`** — expect `ok` and printed shapes `(18,)`.
- [ ] Run the **inline `gym.make`** example above — no exceptions through `reset` + one `step`.

### README / documentation

- [ ] Follow the **ChronoViperWalk** section from a fresh clone with only `PYTHONPATH` set (no undocumented steps).
- [ ] Confirm the **version matrix** rows match what you intend to support (edit table if your CI uses different pins).
- [ ] Skim **Ray + ROCm** section: examples use **`ROCR_VISIBLE_DEVICES`** for AMD GPU visibility (not `CUDA_VISIBLE_DEVICES` as the primary selector) and mention **`ray start --num-gpus=N`**.

### Regression / import hygiene

- [ ] `python3 -c "import gym_chrono"` then `import gymnasium as gym; gym.make('ChronoViperWalk-v0')` on a host **with** PyChrono.
- [ ] Confirm no new **mandatory** dependency on `pychrono.sensor` / `pychrono.vehicle` was introduced for **`ChronoViperWalk-v0`** specifically.

---

## AMD-specific vs portable

| Topic | Scope |
|-------|--------|
| **`ChronoViperWalkEnv` physics** | **Portable CPU** (NSC + Bullet + Viper). Not a HIP kernel port. |
| **`ROCR_VISIBLE_DEVICES`, `HIP_VISIBLE_DEVICES` in Ray `runtime_env`** | **AMD ROCm + Ray** operational guidance for separating **CPU Chrono** workers from **GPU** learners. |
| **`RAY_EXPERIMENTAL_NOSET_*` before `import ray`** | **Ray on PyTorch ROCm** (≥ 2.45) — relevant to AMD ROCm stacks; harmless on many other stacks when set via `setdefault`. |
| **`ray start --num-gpus=N`** | **Ray scheduling** — not vendor-specific, but commonly hits ROCm + MI-class multi-GPU nodes. |

---

## Reviewer reproduction (generic AMD Linux: CPUs + Instinct GPUs)

These steps assume a **64-bit Linux** host with **AMD ROCm** installed for **Instinct** GPUs (e.g. MI200 / MI300 class), **Python 3.10+**, and permission to use the GPU stack. They intentionally avoid **site-specific** paths, schedulers, or cluster-only configuration.

### 0) Sanity checks (optional)

```bash
# ROCm user-land visible
rocm-smi --showproductname || true

# HIP compiler available (build hosts only)
hipconfig --version || true
```

### 1) Build or install **PyChrono** (Chrono 10 + `robot` module)

Point **`CHRONO_DATA_DIR`** at a checkout of Chrono’s **`data/`** tree (textures, meshes). Build PyChrono following the official [Python module installation](https://api.projectchrono.org/module_python_installation.html) guide. Enable at least **Python** and **robot** for `ChronoViperWalk-v0` (other modules optional for this env).

Set:

```bash
export CHRONO_DATA_DIR="/absolute/path/to/chrono/data/"   # trailing slash OK
export PYTHONPATH="/absolute/path/to/chrono/build/bin:${PYTHONPATH:-}"
```

(Adjust `build/bin` to wherever `pychrono` was installed for your build.)

### 2) Check out **this PR branch** of `gym-chrono`

```bash
export GYM_CHRONO_ROOT="${GYM_CHRONO_ROOT:-$HOME/src/gym-chrono}"
git clone https://github.com/amd-pratmish/gym-chrono.git "$GYM_CHRONO_ROOT"
cd "$GYM_CHRONO_ROOT"
git fetch origin feat/chrono10-viper-walk-env
git checkout feat/chrono10-viper-walk-env
export PYTHONPATH="$GYM_CHRONO_ROOT:${PYTHONPATH:-}"
```

(You can instead merge the PR locally or add the fork as `git remote add`.)

### 3) Python venv and **CPU-only** smoke (no Ray, no GPU required for physics)

```bash
python3 -m venv .venv-gym-chrono
source .venv-gym-chrono/bin/activate
python3 -m pip install --upgrade pip
python3 -m pip install "gymnasium>=0.29" "numpy>=1.24,<2"

python3 gym_chrono/test/smoke_viper_walk.py
```

**Expected:** a line starting with `ok` and NumPy shapes `(18,)`.

### 4) Same host, **PyTorch ROCm** + **Ray** (CPU sim actors + GPU visible to driver)

Install a **PyTorch ROCm** build that matches your ROCm driver (see [PyTorch — Get Started](https://pytorch.org/get-started/locally/)), then:

```bash
source .venv-gym-chrono/bin/activate
python3 -m pip install "ray[default]>=2.34"
```

**Single-node Ray** (replace `N` with the number of Instinct GPUs you want Ray to advertise on this host):

```bash
# Select GPUs for the session (AMD Instinct — use ROCR, not CUDA_VISIBLE_DEVICES)
export ROCR_VISIBLE_DEVICES="${ROCR_VISIBLE_DEVICES:-0}"

# Ray + PyTorch ROCm: set before any `import ray` in your driver process
export RAY_EXPERIMENTAL_NOSET_HIP_VISIBLE_DEVICES="${RAY_EXPERIMENTAL_NOSET_HIP_VISIBLE_DEVICES:-1}"
export RAY_EXPERIMENTAL_NOSET_ROCR_VISIBLE_DEVICES="${RAY_EXPERIMENTAL_NOSET_ROCR_VISIBLE_DEVICES:-1}"

ray stop --force 2>/dev/null || true
ray start --head --num-gpus="${NUM_GPUS:-1}" --disable-usage-stats
```

**Driver script** — checked in as **`playground/reviewer_ray_chrono_smoke.py`**. Inline copy for convenience:

```python
#!/usr/bin/env python3
"""Minimal Ray + ChronoViperWalk on AMD: CPU actors, GPU left for torch (optional)."""
from __future__ import annotations

import os

os.environ.setdefault("RAY_EXPERIMENTAL_NOSET_HIP_VISIBLE_DEVICES", "1")
os.environ.setdefault("RAY_EXPERIMENTAL_NOSET_ROCR_VISIBLE_DEVICES", "1")

import numpy as np
import ray

_CHRONO_CPU = {
    "env_vars": {
        "HIP_VISIBLE_DEVICES": "-1",
        "ROCR_VISIBLE_DEVICES": "-1",
    }
}


@ray.remote(num_cpus=1, runtime_env=_CHRONO_CPU)
class ViperActor:
    def __init__(self, seed: int) -> None:
        import gymnasium as gym

        import gym_chrono  # noqa: F401 — registration

        self._env = gym.make("ChronoViperWalk-v0", render_mode=None)
        self._env.reset(seed=seed)

    def step(self) -> float:
        a = self._env.action_space.sample()
        _obs, r, _term, _trunc, _info = self._env.step(a)
        return float(r)


def main() -> None:
    if not os.environ.get("CHRONO_DATA_DIR"):
        raise SystemExit("Set CHRONO_DATA_DIR to your Chrono data directory.")
    ray.init()
    actors = [ViperActor.remote(i) for i in range(2)]
    rewards = ray.get([a.step.remote() for a in actors])
    print("rewards", rewards)
    ray.shutdown()


if __name__ == "__main__":
    main()
```

```bash
chmod +x playground/reviewer_ray_chrono_smoke.py
python3 playground/reviewer_ray_chrono_smoke.py
```

**Optional:** in the same venv, `python3 -c "import torch; print(torch.cuda.is_available(), torch.version.hip)"` should report HIP-enabled PyTorch on a ROCm box. Keep **heavy** GPU work out of the Ray actors that construct `ChronoViperWalk-v0` unless you intentionally port kernels.

### 5) Tear down

```bash
ray stop --force
```

### CODE REVIEW

**Comment by amd-pratmish** [comment]
@harryzhang1018 @DanNegrut
